packages/cli-kit/src/session/device-authorization.ts

```diff
@@ -0,0 +1,106 @@
+import {clientId} from './identity.js'
+import {exchangeDeviceCodeForAccessToken} from './exchange.js'
+import {IdentityToken} from './schema.js'
+import {identity as identityFqdn} from '../environment/fqdn.js'
+import {shopifyFetch} from '../http.js'
+import {content, debug, info, token} from '../output.js'
+
+export interface DeviceAuthorizationResponse {
+  deviceCode: string
+  userCode: string
+  verificationUri: string
+  expiresIn: number
+  verificationUriComplete?: string
+  interval?: number
+}
+
+/**
+ * Initiate a device authorization flow.
+ * This will return a DeviceAuthorizationResponse containing the URL where user
+ * should go to authorize the device without the need of a callback to the CLI.
+ *
+ * Also returns a `deviceCode` used for polling the token endpoint in the next step.
+ *
+ * @param scopes The scopes to request
+ * @returns {Promise<DeviceAuthorizationResponse>} An object with the device authorization response.
+ */
+export async function requestDeviceAuthorization(scopes: string[]): Promise<DeviceAuthorizationResponse> {
+  const fqdn = await identityFqdn()
+  const identityClientId = await clientId()
+  const queryParams = {client_id: identityClientId, scope: scopes.join(' ')}
+  const url = `https://${fqdn}/oauth/device_authorization`
+
+  const response = await shopifyFetch('identity', url, {
+    method: 'POST',
+    headers: {'Content-type': 'application/x-www-form-urlencoded'},
+    body: convertRequestToParams(queryParams),
+  })
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const jsonResult: any = await response.json()
+
+  debug(content`Received device authorization code: ${token.json(jsonResult)}`)
+
+  info('\nTo run this command, log in to Shopify Partners.')
+  info(content`User verification code: ${jsonResult.user_code}`)
+  info(content`👉 Open ${token.link('this Link', jsonResult.verification_uri_complete)} to start the auth process`)
```

Comment:
What about showing the full URL instead of the hyperlink here? Users will have to copy&paste it sometimes instead of just clicking, right?

Reply:
I'm worried about other environments where a browser can't be opened automatically, but we can change it later.